src/StorageFlattening.cpp

```diff
@@ -202,8 +285,28 @@ class FlattenDimensions : public IRMutator {
         }
         stmt = LetStmt::make(op->name + ".buffer", builder.build(), stmt);
 
-        // Make the allocation node
-        stmt = Allocate::make(op->name, op->types[0], op->memory_type, allocation_extents, condition, stmt);
+        if (hoisted_storages_map.count(op->name) > 0) {
+            HoistedStorageData &hoisted_storage_data = hoisted_storages[hoisted_storages_map[op->name]];
+            vector<Expr> bounded_extents;
+            for (const auto &e : allocation_extents) {
+                Expr expanded_extent = e;
+                // Iterate from innermost outwards
+                for (auto it = hoisted_storages.rbegin(); it != hoisted_storages.rend(); it++) {
+                    expanded_extent = expand_expr(expanded_extent, it->scope);
+                }
+                expanded_extent = simplify(substitute_in_all_lets(expanded_extent));
```

Comment:
simplify(substitute_in_all_lets(x)) is not safe. substitute_in_all_lets creates a DAG which could be combinatorially large when viewed as a tree, and you can only really pass such an Expr to an IRGraphMutator/Visitor or to cse

Would simplify(common_subexpression_elimination(expanded_extent)) work just as well?

Reply:
Done.